snow/engine/snowman/transitive.go

```diff
@@ -478,41 +483,79 @@ func (t *Transitive) Start(ctx context.Context, startReqID uint32) error {
 		return err
 	}
 
+	var (
+		preferredID    ids.ID
+		preferredBlock snowman.Block
+	)
+
+	// Default to using the last accepted block if the VM does not define its
+	// own initial preferred block.
+	if t.getInitialPreferenceVM == nil {
+		preferredID = lastAcceptedID
+		preferredBlock = lastAccepted
+	} else {
+		preferredID, err = t.getInitialPreferenceVM.GetInitialPreference(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get initial preference: %w", err)
+		}
+
+		preferredBlock, err = t.VM.GetBlock(ctx, preferredID)
+		if err != nil {
+			return fmt.Errorf("failed to get initial preferred block: %w", err)
+		}
+	}
+
+	// Check if the initial setup is more recent than the last accepted
+	// block. This can occur if we shut down while a set of blocks was still
+	// processing in consensus.
+	preferredTip := lastAcceptedID
+	if preferredBlock.Height() > lastAccepted.Height() {
+		preferredTip = preferredID
+	}
+
+	preferredChain := make([]snowman.Block, 0)
+	for preferredBlock.Height() > lastAccepted.Height() {
+		preferredChain = append(preferredChain, preferredBlock)
+		preferredBlock, err = t.VM.GetBlock(ctx, preferredBlock.Parent())
+		if err != nil {
+			return err
+		}
+	}
+
 	// initialize consensus to the last accepted blockID
 	lastAcceptedHeight := lastAccepted.Height()
 	if err := t.Consensus.Initialize(t.Ctx, t.Params, lastAcceptedID, lastAcceptedHeight, lastAccepted.Timestamp()); err != nil {
 		return err
 	}
 
-	// to maintain the invariant that oracle blocks are issued in the correct
-	// preferences, we need to handle the case that we are bootstrapping into an oracle block
-	if oracleBlk, ok := lastAccepted.(snowman.OracleBlock); ok {
-		options, err := oracleBlk.Options(ctx)
-		switch {
-		case err == snowman.ErrNotOracle:
-			// if there aren't blocks we need to deliver on startup, we need to set
-			// the preference to the last accepted block
-			if err := t.VM.SetPreference(ctx, lastAcceptedID); err != nil {
+	issuedMetric := t.metrics.issued.WithLabelValues(builtSource)
+
+	// Re-issue all blocks in the preferred chain into consensus if the
+	// preferred chain extends the last accepted block
+	if len(preferredChain) > 0 && preferredChain[len(preferredChain)-1].Parent() == lastAcceptedID {
+		// reverse the preferred chain so the blocks are now in chronological order
+		slices.Reverse(preferredChain)
+
+		for _, preferredBlk := range preferredChain {
+			if _, err := t.addUnverifiedBlockToConsensus(
```

Comment:
I think returning `false` here should either FATAL the node... or be handled... But ignoring it seems problematic.

If we previously persisted a block as preferred and now we are attempting to verify it and it fails verification, we either need to stop inserting the preferred chain or we need to die.

I think dying may be problematic... So it seems like we need to abandon the remaining preferred chain.

Reply:
Updated + added test